apis/gateway/v1beta2/apirule_types.go

```diff
@@ -0,0 +1,236 @@
+type APIRule struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   APIRuleSpec   `json:"spec,omitempty"`
+	Status APIRuleStatus `json:"status,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// APIRuleList contains a list of ApiRule
+type APIRuleList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []APIRule `json:"items"`
+}
+
+// Service .
+type Service struct {
+	// Specifies the name of the exposed service.
+	Name *string `json:"name"`
+	// Specifies the Namespace of the exposed service. If not defined, it defaults to the APIRule Namespace.
+	// +kubebuilder:validation:Pattern=^[a-z0-9]([-a-z0-9]*[a-z0-9])?$
+	// +optional
+	Namespace *string `json:"namespace,omitempty"`
+	// Specifies the communication port of the exposed service.
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:validation:Maximum=65535
+	Port *uint32 `json:"port"`
+	// Specifies if the service is internal (in cluster) or external.
+	// +optional
+	IsExternal *bool `json:"external,omitempty"`
+}
+
+// Rule .
+type Rule struct {
+	// Specifies the path of the exposed service.
+	// +kubebuilder:validation:Pattern=^([0-9a-zA-Z./*()?!\\_-]+)
+	Path string `json:"path"`
+	// Describes the service to expose. Overwrites the **spec** level service if defined.
+	// +optional
+	Service *Service `json:"service,omitempty"`
+	// Represents the list of allowed HTTP request methods available for the **spec.rules.path**.
+	// +kubebuilder:validation:MinItems=1
+	Methods []HttpMethod `json:"methods"`
+	// Disables authorization when set to true.
+	// +optional
+	NoAuth *bool `json:"noAuth"`
+	// Specifies the Istio JWT access strategy.
+	// +optional
+	Jwt *JwtConfig `json:"jwt,omitempty"`
+	// Specifies the list of [Ory Oathkeeper](https://www.ory.sh/docs/oathkeeper/pipeline/mutator) mutators.
+	// +optional
+	Mutators []*Mutator `json:"mutators,omitempty"`
+	// +optional
+	Timeout *Timeout `json:"timeout,omitempty"`
+}
+
+// HttpMethod specifies the HTTP request method. The list of supported methods is defined in RFC 9910: HTTP Semantics and RFC 5789: PATCH Method for HTTP.
+// +kubebuilder:validation:Enum=GET;HEAD;POST;PUT;DELETE;CONNECT;OPTIONS;TRACE;PATCH
+type HttpMethod string
+
+// APIRuleResourceStatus describes the status of APIRule.
+type APIRuleResourceStatus struct {
+	Code        StatusCode `json:"code,omitempty"`
+	Description string     `json:"desc,omitempty"`
+}
+
+func init() {
+	SchemeBuilder.Register(&APIRule{}, &APIRuleList{})
+}
+
+// Mutator is a configuration for Istio mutators. It is used to enrich an incoming request with information.
+type Mutator struct {
+	// Specifies the name of the mutator.
+	Handler string `json:"handler"`
+	// Configures the mutator. Configuration keys vary per mutator.
+	// +kubebuilder:validation:Type=object
+	// +kubebuilder:pruning:PreserveUnknownFields
+	Config *runtime.RawExtension `json:"config,omitempty"`
```

Comment:
I think we missed to think about improving typing for mutators. Maybe this is something we want to discuss in a follow up?

Reply:
yes, this makes sense, I do not like the `*runtime.RawExtension` approach as well. @werdes72 would you check if we can have it structured for mutators and also propose update in ADR ?

Reply:
Added more info to the mutators task, for now let's keep it.